### PR TITLE
CSHARP-4719: Benchmark performance of in-use Encryption

### PR DESCRIPTION
**Pull request by adelinowona**
Benchmark implemented per this spec: https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/benchmarks.rst 

Important to note that I am using the pure time metrics for running a decrypt operation as performance data rather than counting the number of decrypt operations we can complete in a second. I am doing this to keep uniformity in the benchmarking mechanism for our driver (using benchmarkDotNet) and also using the time metrics seems enough to me to see any performance gains. 

Here is the data from benchmarks showing improvements made.

Benchmarks without improvements:
| Method                     | ThreadsCount | Mean       | Error      | StdDev     |
|--------------------------- |------------- |-----------:|-----------:|-----------:|
| BulkDecryptionUsingBinding | 1            |   7.563 ms |  0.4013 ms |  0.3753 ms |
| BulkDecryptionUsingBinding | 2            |  10.876 ms |  1.1460 ms |  1.0719 ms |
| BulkDecryptionUsingBinding | 8            |  84.420 ms |  2.4353 ms |  2.2780 ms |
| BulkDecryptionUsingBinding | 64           | 878.740 ms | 37.7653 ms | 35.3257 ms |

------------------------------------------------------------------------------------------------

Benchmarks with improvements:
| Method                     | ThreadsCount | Mean       | Error      | StdDev     |
|--------------------------- |------------- |-----------:|-----------:|-----------:|
| BulkDecryptionUsingBinding | 1            |   6.181 ms |  0.3903 ms |  0.3650 ms |
| BulkDecryptionUsingBinding | 2            |   9.630 ms |  0.4241 ms |  0.3760 ms |
| BulkDecryptionUsingBinding | 8            |  80.497 ms |  4.2131 ms |  3.9409 ms |
| BulkDecryptionUsingBinding | 64           | 714.617 ms | 58.1227 ms | 54.3680 ms |

We get modest improvements but still, the most impactful change would be to enable usage of the native libmonogcrypt crypto implementation as outlined in [DRIVERS-2718](https://jira.mongodb.org/browse/DRIVERS-2718).

### CODE REVIEW

File: CSharpDriver.sln [hunk 1]
```diff
@@ -64,6 +64,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "SmokeTests", "SmokeTests",
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MongoDB.TestHelpers", "tests\MongoDB.TestHelpers\MongoDB.TestHelpers.csproj", "{DF888021-744F-4A8B-9324-831DEFC48AB8}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BenchmarkLibmongocrypt", "BenchmarkLibmongocrypt\BenchmarkLibmongocrypt.csproj", "{F2C683ED-138E-48AF-9C49-2BF8A0631842}"
```
Comment (adelinowona) [inline CSharpDriver.sln:67]: Once the driver benchmarking PR is merged, I plan on placing this encryption benchmarks under the new benchmarks solution folder  introduced in the driver benchmarking PR.

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using MongoDB.Bson;
+using MongoDB.Driver;
+using System.Threading;
+using System.Reflection;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Columns;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Reports;
```
Comment (BorisDog) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:9]: Remove and sort usings.

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+        private const int RepeatCount = 10;
+
+        [Params(1, 2, 8, 64)]
+        public int ThreadCounts;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            byte[] localMasterKey = Convert.FromBase64String(LocalMasterKey);
```
Comment (BorisDog) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:36]: var

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        [Benchmark]
+        public void BulkDecryptionUsingBinding()
+        {
+            var tasks = new Task[ThreadCounts];
+            for (int i = 0; i < ThreadCounts; i++)
+            {
+                tasks[i] = Task.Factory.StartNew(BulkDecryptionTaskUsingBinding());
+            }
+
+            Task.WaitAll(tasks);
```
Comment (BorisDog) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:128]: We need to use real threads, or `ThreadPerTaskScheduler`.
Also benchmark measures thread creations, so maybe it's worth creating the threads in GlobalThread.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+            _encryptedValuesDocumentBytes = encryptedValuesDocument.ToBson();
+
+            _unencryptedClientCollection = new MongoClient().GetDatabase("crypt-test")
+                .GetCollection<BsonDocument>("encryptedValues");
+
+            _encryptedClientCollection = keyVaultClient.GetDatabase("crypt-test").GetCollection<BsonDocument>("encryptedValues");
+            _encryptedClientCollection.InsertOne(encryptedValuesDocument);
+
+            // since the benchmark is not in the same assembly as MongoClient, we can't directly get the LibMongoCryptController in it
```
Comment (sanych-sun) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:89]: Can InternalsVisibleTo attribute help with this like we doing for unit-tests? Reflection is known to be a little slower.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+            _unencryptedClientCollection = new MongoClient().GetDatabase("crypt-test")
+                .GetCollection<BsonDocument>("encryptedValues");
+
+            _encryptedClientCollection = keyVaultClient.GetDatabase("crypt-test").GetCollection<BsonDocument>("encryptedValues");
+            _encryptedClientCollection.InsertOne(encryptedValuesDocument);
+        }
+
+        [Benchmark(Baseline = true)]
+        public void FindWithNoEncryption()
```
Comment (adelinowona) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:98]: Will remove this

File: MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                for (int i = 0; i < RepeatCount; i++)
+                {
+                    _unencryptedClientCollection.Find(Builders<BsonDocument>.Filter.Empty).Single();
+                }
+            });
+        }
+
+        [Benchmark]
+        public void FindWithEncryption()
```
Comment (adelinowona) [inline MongoDB.LibMongoCryptBindingsBenchmark/Benchmarks.cs:110]: Will remove this as well

File: benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        private DisposableMongoClient _disposableKeyVaultClient;
+        private AutoEncryptionLibMongoCryptController _libMongoCryptController;
+
+        [Params(1, 2, 8, 64)]
+        public int ThreadsCount;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            byte[] localMasterKey = Convert.FromBase64String(LocalMasterKey);
```
Comment (BorisDog) [inline benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs:45]: nit: var

File: benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                    var toEncryptString = $"value {(i + 1):D4}";
+                    var encryptedString =
+                        clientEncryption.Encrypt(toEncryptString, encryptOptions, CancellationToken.None);
+                    encryptedValuesDocument.Add(new BsonElement($"key{(i + 1):D4}", encryptedString));
+                }
+
+                _encryptedValuesDocumentBytes = encryptedValuesDocument.ToBson();
+
+                // Create libmongocrypt binding that will be used for decryption
+                var cryptClient = CryptClientCreator.CreateCryptClient(autoEncryptionOptions.ToCryptClientSettings());
```
Comment (BorisDog) [inline benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs:94]: I think `cryptClient` needs to be disposed as well.

**Review by BorisDog — CHANGES_REQUESTED** [review]

**Review by sanych-sun — APPROVED** [review]
LGTM

File: benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                _libMongoCryptController =
+                    AutoEncryptionLibMongoCryptController.Create(_disposableKeyVaultClient, cryptClient, autoEncryptionOptions);
+            }
+        }
+
+        [Benchmark]
```
Comment (sanych-sun) [inline benchmarks/MongoDB.Libmongocrypt.Benchmarks/LibmongocryptBindingBenchmark.cs:100]: I can see only one benchmark here, but PR description refers to 3. Could you please describe in readme or somewhere else how to perform such benchmarking

File: benchmarks/MongoDB.Driver.Benchmarks/LibmongocryptBindingBenchmark.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+using MongoDB.Driver.TestHelpers;
+using MongoDB.Libmongocrypt;
+
+namespace MongoDB.Benchmarks
+{
+    public class LibmongocryptBindingBenchmark
+    {
+        private const int RepeatCount = 10;
+        private const string LocalMasterKey =
+            "Mng0NCt4ZHVUYUJCa1kxNkVyNUR1QURhZ2h2UzR2d2RrZzh0cFBwM3R6NmdWMDFBMUN3YkQ5aXRRMkhGRGdQV09wOGVNYUMxT2k3NjZKelhaQmRCZGJkTXVyZG9uSjFk";
```
Comment (BorisDog) [inline benchmarks/MongoDB.Driver.Benchmarks/LibmongocryptBindingBenchmark.cs:34]: Same line is fine here.

File: benchmarks/MongoDB.Driver.Benchmarks/LibmongocryptBindingBenchmark.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                    keyId: dataKeyId);
+
+                var encryptedValuesDocument = new BsonDocument();
+                for (int i = 0; i < 1500; i++)
+                {
+                    var toEncryptString = $"value {(i + 1):D4}";
+                    var encryptedString =
+                        clientEncryption.Encrypt(toEncryptString, encryptOptions, CancellationToken.None);
+                    encryptedValuesDocument.Add(new BsonElement($"key{(i + 1):D4}", encryptedString));
+                }
```
Comment (BorisDog) [inline benchmarks/MongoDB.Driver.Benchmarks/LibmongocryptBindingBenchmark.cs:91]: minor: Looks like the scope of `using clientEncryption` can end here.

File: benchmarks/MongoDB.Driver.Benchmarks/BenchmarkRunner.cs [hunk 13]
```diff
@@ -24,27 +24,30 @@ public class BenchmarkRunner
...
-                .WithOption(ConfigOptions.JoinSummary, true)
-                .AddExporter(new LocalExporter())
-                .HideColumns("BenchmarkDataSetSize");
+            var config = DefaultConfig.Instance;
 
-            if (exportingToEvergreen)
-            {
-                config = config.AddExporter(new EvergreenExporter(evergreenOutputFile));
-            }
+            config = executingDriverBenchmarks
```
Comment (BorisDog) [inline benchmarks/MongoDB.Driver.Benchmarks/BenchmarkRunner.cs:43]: nit: IMHO having regular assignments here are more readable, as u don't need to validate the `else` no-op clause:

```
if (..)
{
 config = config.With(...)
}
```

**Review by BorisDog — APPROVED** [review]
LGTM (few minor comments)

**Pull request merged**